libraries/AP_RangeFinder/AP_RangeFinder_NMEA.cpp

```diff
@@ -0,0 +1,188 @@
+    if (uart != nullptr) {
+        uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance));
+    }
+}
+
+// detect if a NMEA rangefinder by looking to see if the user has configured it
+bool AP_RangeFinder_NMEA::detect(AP_SerialManager &serial_manager, uint8_t serial_instance)
+{
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance) != nullptr;
+}
+
+// update the state of the sensor
+void AP_RangeFinder_NMEA::update(void)
+{
+    uint32_t now = AP_HAL::millis();
+    if (get_reading(state.distance_cm)) {
+        // update range_valid state based on distance measured
+        _last_reading_ms = now;
+        update_status();
+    } else if ((now - _last_reading_ms) > 3000) {
+        set_status(RangeFinder::RangeFinder_NoData);
+    }
+}
+
+// return last value measured by sensor
+bool AP_RangeFinder_NMEA::get_reading(uint16_t &reading_cm)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    // read any available lines from the lidar
+    float sum = 0.0f;
+    uint16_t count = 0;
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        char c = uart->read();
+        if (decode(c)) {
+            sum += _distance_m;
+            count++;
+        }
+    }
+
+    // return false on failure
+    if (count == 0) {
+        return false;
+    }
+
+    // return average of all measurements
+    reading_cm = 100.0f * sum / count;
+    return true;
+}
+
+// add a single character to the buffer and attempt to decode
+// returns true if a complete sentence was successfully decoded
+bool AP_RangeFinder_NMEA::decode(char c)
+{
+    switch (c) {
+    case ',':
+        // end of a term, add to checksum
+        _checksum ^= c;
+        FALLTHROUGH;
+    case '\r':
+    case '\n':
+    case '*':
+    {
+        bool valid_sentence = false;
+        if (_term_offset < sizeof(_term)) {
+            _term[_term_offset] = 0;
+            valid_sentence = decode_latest_term();
+        }
+        _term_number++;
+        _term_offset = 0;
+        _term_is_checksum = (c == '*');
+        return valid_sentence;
+    }
+
+    case '$': // sentence begin
+        _sentence_type = SONAR_UNKNOWN;
+        _term_number = 0;
+        _term_offset = 0;
+        _checksum = 0;
+        _term_is_checksum = false;
+        _distance_m = -1.0f;
+        return false;
+    }
+
+    // ordinary characters are added to term
+    if (_term_offset < sizeof(_term) - 1)
+        _term[_term_offset++] = c;
+    if (!_term_is_checksum)
+        _checksum ^= c;
+
+    return false;
+}
+
+// decode the most recently consumed term
+// returns true if new sentence has just passed checksum test and is validated
+bool AP_RangeFinder_NMEA::decode_latest_term()
+{
+    // handle the last term in a message
+    if (_term_is_checksum) {
+        uint8_t checksum = 16 * char_to_hex(_term[0]) + char_to_hex(_term[1]);
```

Comment:
I poked around with strtoul and, as you suggest, it works.  Apparently peter and tridge had a discussion about this and Tridge suggested we should have our own decoder so I've left it as is.  It's possible I've taken this comment out of context but because you don't seem super insistent on this I'll leave it as is.  txs again.